File: Assets/Disease Destroyer Assets/Scripts/playerMovement.js
private var maxX = 490;
private var minX = -490;
private var maxY = 490;
private var minY = -490;
var speedSlowFactor=1;
var accelSlowFactor =0;
var slowDuration=3;
var slowTime=0;

public var slow=false;
var speed : float;
var bullet : Rigidbody;
var bulletSpeed : float;
var clone : Transform;
var burstLength : float;
var burstAngle : float;
var burstVel : float;
var burstDuration : float;
var shoot  : AudioClip;//shoot sound
var push : AudioClip;//push sound
private var burstTime = 0.0;
private var startTime;
var curBullet: Rigidbody;
// Use this for initialization
function Start () {
	startTime=Time.time;
}
function restart(){
	rigidbody.position=Vector3.zero;
	rigidbody.velocity=Vector3.zero;
	startTime=Time.time;
}
//Update is called once per frame
function Update(){
	//Finds number of bullets, fires if there are none
	var numBullets = GameObject.FindGameObjectsWithTag("Bullet").Length;
	if (Time.timeScale!=0&&Input.GetButtonDown("Fire1")) {
		if(numBullets == 0){
		    AudioSource.PlayClipAtPoint(shoot, transform.position);
			curBullet = Instantiate(bullet, GameObject.Find("BulletSpawn").transform.position, Quaternion(0, 0, 0, 0));
			curBullet.velocity = Vector3.Normalize(GameObject.Find("BulletSpawn").transform.position - transform.position) * bulletSpeed;// + GameObject.Find("Character").rigidbody.velocity;
		}
		else{
			curBullet.GetComponent(bulletControl).kill();
		}
		 // Instantiate the projectile at the position and rotation of this transform
		 
		
	}	
	if(Time.timeScale!=0&&(Input.GetButtonDown("Fire2") || Input.GetKeyDown("space")) && burstTime <= 0){
		AudioSource.PlayClipAtPoint(push, transform.position);
		burstTime = burstDuration;
		GameObject.Find("PushLight").GetComponent("pushLightScript").push();
		burstAttack();
	}
	else{
		burstTime = burstTime - Time.deltaTime;
	}
}

// FixedUpdate is called with physics
function FixedUpdate () {
	if((Time.time>(this.startTime+1))){//only move if the game isn't over
		PlayerFacing();
    	if(rigidbody.velocity.magnitude<350){
    		rigidbody.velocity.y =  rigidbody.velocity.y+(Input.GetAxis("Vertical") * (speed-accelSlowFactor));
    		rigidbody.velocity.x =  rigidbody.velocity.x+(Input.GetAxis("Horizontal") * (speed-accelSlowFactor));
    	}	
    }
    else{
    	rigidbody.position=Vector3.zero;
		rigidbody.velocity=Vector3.zero;
    }
    if(rigidbody.transform.position.x>maxX){//if the virus leaves the bounding box bring it back in and send it toward the center
    	rigidbody.transform.position.x=maxX;
    	rigidbody.velocity.x=-Mathf.Abs(rigidbody.velocity.x*.5);
    }
    if(rigidbody.transform.position.x<minX){
    	rigidbody.transform.position.x=minX;
    	rigidbody.velocity.x=Mathf.Abs(rigidbody.velocity.x*.5);
    }
    if(rigidbody.transform.position.y>maxY){
    	rigidbody.transform.position.y=maxY;
    	rigidbody.velocity.y=-Mathf.Abs(rigidbody.velocity.y*.5);
    }
    if(rigidbody.transform.position.y<minY){
    	rigidbody.transform.position.y=minY;
    	rigidbody.velocity.y=Mathf.Abs(rigidbody.velocity.y*.5);
    }
	if(rigidbody.velocity.magnitude>=450){
		rigidbody.velocity=rigidbody.velocity*(.95*speedSlowFactor);
	}
	else{
		rigidbody.velocity=rigidbody.velocity*(.99*speedSlowFactor);
	}
	endSlow();
}
function addVel(velocity: Vector3){
	rigidbody.velocity.x=velocity.x;
	rigidbody.velocity.y=velocity.y;	
}

function startSlow(){
	if(!slow){
		slow=true;
		slowTime=Time.time;
		speedSlowFactor=1;
		accelSlowFactor=3;
	}
}
function endSlow(){
	if(slow && Time.time>this.slowTime+slowDuration){
		slowTime=0;
		slow=false;
		speedSlowFactor=1;
		accelSlowFactor=0;
	}
}

function burstAttack(){
	yield WaitForSeconds(burstDuration / 2);
	var colliders = Physics.OverlapSphere(transform.position, burstLength);
	for(var hit : Collider in colliders){
		if(hit.rigidbody){
			//Detects whether they're within the proper angle
			var angle = Vector3.Angle(hit.rigidbody.position - transform.position, transform.forward);
			if(angle <= burstAngle){
				//BURST CODE HERE
				var burstDir = Vector3.Normalize(hit.rigidbody.position - transform.position);
				hit.rigidbody.velocity = hit.rigidbody.velocity + burstDir * burstVel;
			}
		}
	}
}

//Player always faces the mouse, does some voodoo shit with raycasting to determine direction to face
function PlayerFacing(){
	var hitdist = 1.0;
	var ray = Camera.main.ScreenPointToRay(Input.mousePosition);
	var playerPlane = new Plane(Vector3(0,0,1), transform.position);
	if(playerPlane.Raycast(ray, hitdist)){
		var targetPoint = ray.GetPoint(hitdist);
		var rotation = Quaternion.LookRotation(targetPoint - transform.position);
		//transform.rotation = Quaternion.Slerp(transform.rotation, rotation, Time.deltaTime);
		transform.rotation = rotation;
	}
}

File: Assets/Disease Destroyer Assets/Scripts/playerMovement.cs
using System.Collections;
using UnityEngine;

[RequireComponent(typeof(Rigidbody))]
public class playerMovement : MonoBehaviour
{
    public float speedSlowFactor=1, accelSlowFactor, slowDuration=3, slowTime;
    public bool slow;
    public float speed, bulletSpeed, burstLength, burstAngle, burstVel, burstDuration;
    public Rigidbody bullet, curBullet;
    public Transform clone;
    public AudioClip shoot, push;
    float burstTime, startTime;
    Rigidbody body; Collider playerCollider;
    void Awake() { body=GetComponent<Rigidbody>();playerCollider=GetComponentInChildren<Collider>(); }
    void Start() => startTime = Time.time;
    public void restart() { body.position=Vector3.zero; body.linearVelocity=Vector3.zero; startTime=Time.time; }
    void Update()
    {
        if (Time.timeScale != 0 && Input.GetButtonDown("Fire1"))
        {
            if (!curBullet) { HUDHandler.PlaySfx(shoot); var spawn=GameObject.Find("BulletSpawn").transform; curBullet=Instantiate(bullet,spawn.position,Quaternion.identity); curBullet.linearVelocity=(spawn.position-transform.position).normalized*bulletSpeed; }
            else curBullet.GetComponent<bulletControl>().kill();
        }
        if (Time.timeScale != 0 && (Input.GetButtonDown("Fire2") || Input.GetKeyDown(KeyCode.Space)) && burstTime <= 0) { HUDHandler.PlaySfx(push); burstTime=burstDuration; var light=GameObject.Find("PushLight"); if(light) light.GetComponent<pushLightScript>()?.push(); StartCoroutine(BurstAttack()); }
        else burstTime -= Time.deltaTime;
    }
    void FixedUpdate()
    {
        if (Time.time > startTime+1) { FacePointer(); if(body.linearVelocity.magnitude<350){body.linearVelocity += new Vector3(Input.GetAxis("Horizontal")*(speed-accelSlowFactor),Input.GetAxis("Vertical")*(speed-accelSlowFactor),0);} }
        else {body.position=Vector3.zero;body.linearVelocity=Vector3.zero;}
        var p=body.position; if(p.x>490){p.x=490;body.linearVelocity=new Vector3(-Mathf.Abs(body.linearVelocity.x*.5f),body.linearVelocity.y,0);} if(p.x<-490){p.x=-490;body.linearVelocity=new Vector3(Mathf.Abs(body.linearVelocity.x*.5f),body.linearVelocity.y,0);} if(p.y>490){p.y=490;body.linearVelocity=new Vector3(body.linearVelocity.x,-Mathf.Abs(body.linearVelocity.y*.5f),0);} if(p.y<-490){p.y=-490;body.linearVelocity=new Vector3(body.linearVelocity.x,Mathf.Abs(body.linearVelocity.y*.5f),0);} body.position=p;
        body.linearVelocity*=body.linearVelocity.magnitude>=450?.95f*speedSlowFactor:.99f*speedSlowFactor; SeparateFromCells(); EndSlow();
    }
    void SeparateFromCells()
    {
        if(!playerCollider)return;
        var bounds=playerCollider.bounds;
        foreach(var other in Physics.OverlapBox(bounds.center,bounds.extents,Quaternion.identity,~0,QueryTriggerInteraction.Ignore))
        {
            if(other.attachedRigidbody==body)continue;
            if(!other.GetComponentInParent<cellMovement>()&&!other.GetComponentInParent<whiteCellMovement>())continue;
            if(Physics.ComputePenetration(playerCollider,playerCollider.transform.position,playerCollider.transform.rotation,other,other.transform.position,other.transform.rotation,out var direction,out var distance))
            {
                var correction=direction*(distance+.1f);
                if(Time.time<=startTime+1&&other.attachedRigidbody)other.attachedRigidbody.position-=correction;
                else body.position+=correction;
            }
        }
        var position=body.position;position.z=0;body.position=position;
    }
    public void addVel(Vector3 v)=>body.linearVelocity=new Vector3(v.x,v.y,0);
    public void startSlow(){if(!slow){slow=true;slowTime=Time.time;speedSlowFactor=1;accelSlowFactor=3;}}
    void EndSlow(){if(slow&&Time.time>slowTime+slowDuration){slow=false;slowTime=0;speedSlowFactor=1;accelSlowFactor=0;}}
    IEnumerator BurstAttack(){yield return new WaitForSeconds(burstDuration/2);foreach(var hit in Physics.OverlapSphere(transform.position,burstLength)){if(hit.attachedRigidbody&&Vector3.Angle(hit.attachedRigidbody.position-transform.position,transform.forward)<=burstAngle)hit.attachedRigidbody.linearVelocity+=(hit.attachedRigidbody.position-transform.position).normalized*burstVel;}}
    void FacePointer(){var cam=Camera.main;if(!cam)return;var ray=cam.ScreenPointToRay(Input.mousePosition);var plane=new Plane(Vector3.forward,transform.position);if(plane.Raycast(ray,out var d))transform.rotation=Quaternion.LookRotation(ray.GetPoint(d)-transform.position);}
}
